### PR TITLE
Added PID Profiles to Programming framework

### LINKED ISSUE #5050: Multiple PID profiles on switch
## Current Behavior
Currently, using the CLI, it is possible to set multiple profiles in iNAV (https://github.com/iNavFlight/inav/blob/master/docs/Profiles.md). However, I haven't found a way to also set multiple PID profiles and use a RC switch to change between them.

In Betaflight, I believe it is possible to set multiple PID profiles and use a switch to change between them. (https://www.reddit.com/r/Multicopter/comments/66tmdo/can_i_assign_2_different_pid_configurations_in/). 

## Desired Behavior
Add the option to iNAV (either through CLI or Configurator) to switch between multiple PID profiles, similarly to how it is currently already implemented in BetaFlight. 

## Suggested Solution
Would it be possible to fork the existing code regarding switching between multiple PID profiles from Betaflight to iNAV? 

## Who does this impact? Who is this for?
This helps users who need to switch between multiple PID profiles during or between flights (e.g. aircraft with multiple means of propulsion like VTOL, people that want to fly with different types of batteries, people that want to change payloads and CG in between flights, etc.

### LINKED ISSUE #5416: Profile switching via RC adjustments non functional
## Current Behavior
Currently rate profile switching via RC adjustments is not implemented in INAV

## Steps to Reproduce
1. Setup Rate profile switching via a 3 pos. switch in Adjustments tab
2. Change switch positions - rate profile is not changed.

## Expected behavior
1. Rate profile is changed depending on the switch position on the radio transmitter.

## Suggested solution(s)
Fix rate profile switching via RC adjustments function.
OR
Remove this option from INAV configurator and INAV manuals.

## Additional context
Rate switching currently is working in Betaflight and was working in Cleanflight but has never been enabled in INAV, yet it's still present in the latest INAV configurator and described in the manuals.

- INAV version string: 1.6-2.4

### LINKED ISSUE #6631: Please add ability to switch profiles with switch
Hi, can you please add ability to switch profiles with for switch (3 pos switch)? In adjustments there is option - change rateprofile, but it doesn´t work. I was used to this option in betaflight and now it is the only thing that I am missing.
Thanks. Regards Jakub.

### LINKED ISSUE #7322: Rate profile selection on a switch
## Current Behavior

I cannot make the adjustments tab change rate profiles based on an RC channel position.

## Desired Behavior

Setting an RC channel to a certain range gives a change in rate profile.

## Suggested Solution

Allow up to three rate profiles to be selected according to RC channel position.

## Who does this impact? Who is this for?

Many users. It would allow dual/triple rates or PID profiles on a switch. Useful for tuning too.

## Additional context

Some aircraft such as my full house glider have potentially different roll rates depending on whether full span ailerons are deployed. It is not always desirable to have full span ailerons active as they add to drag when thermalling. Tying the aileron/flap mode switch to separate rate profiles would be handy to allow a rate/PID profile for each mode. (Eg thermal mode with ailerons only and no rudder stabilisation, then aerobatic mode using full span ailerons and increased rudder P and I gains.)

### PR DESCRIPTION
**Pull request by MrD-RC**
## Added PID Profiles to Programming framework
This allows setting the profile and getting the active profile in the programming framework.

Fixes #7322
Fixes #6631
Fixes #5416
Fixes #5050
Fixes https://github.com/iNavFlight/inav/projects/4#card-36897432

### Requirements
This change requires configurator change https://github.com/iNavFlight/inav-configurator/pull/1340

### Media

image

Video showing the change working. 
Note: The OSD output in not in the PR. I just hacked the VTx Power to show the profile.
https://youtu.be/7jE-TaM7b4Y

### CODE REVIEW

File: src/main/programming/logic_condition.c [hunk 1]
```diff
@@ -333,6 +334,15 @@ static int logicConditionCompute(
             }
             break;
 
+        case LOGIC_CONDITION_SET_PID_PROFILE:
+            operandA--;
+            if ( getConfigProfile() != operandA  && (operandA >= 0 && operandA < MAX_PROFILE_COUNT)) {
+                return setConfigProfile(operandA);
```
Comment (DzikuVx) [inline src/main/programming/logic_condition.c]: This is not enough. I mean, yes, it activates the profile, but does not switches PID controller to the new one. See for example `pidInitFilters` and `pidInit`

**Review by DzikuVx — CHANGES_REQUESTED** [review]

File: src/main/programming/logic_condition.c (hunk 1 above)
Comment (MrD-RC) [reply src/main/programming/logic_condition.c]: I based this on the CMS. But I guess that does a save and reboot, so none of that is necessary. I'll get it updated.

**Comment by MrD-RC** [comment]
I've made some changes which I hope to test in flight tomorrow. But here is some progress to report, as all seems to be changing  now.

Here is a video showing the profiles changing on the OSD, as well as in configurator.
https://youtu.be/8Q4Rva7XUZc

Here is a video showing the profiles changing in an active model on the ground.
https://youtu.be/vixWCaBF3D0

Here is the programming used for the switch in the above video
image

here are the PID profiles used in the above video
iNav - Profile switch in Programming Framework - PIDs - Profile 1
iNav - Profile switch in Programming Framework - PIDs - Profile 2

One other thing that I think I will add to this PR is an OSD element showing the profile number. Currently I'm displaying it in my OSD by setting it as a Global Variable. However, if people want to have multiple PID/Rate profiles that can be switched in flight. I'm sure they will want to know which profile is in use, especially on a tuning OSD.

Is everyone happy with the naming on this too. I've used PID Profile. But Config profile may be more suitable? Or even just profile? After all, this effects PIDs, rates, filters,  It would be great to get some feedback on that. What are your thoughts @DzikuVx

**Comment by avsaase** [comment]
Great work! I think people who fly motored gliders will be very happy with this. 

Does the PID profile affect loiter settings?

**Comment by MrD-RC** [comment]
@avsaase only loiter direction for some reason. It would make more sense to have loiter radius in there. As with different rates and maximum angles, the loiter radius could be effected.

**Comment by avsaase** [comment]
For builds where speeds is strongly correlated with throttle TPA works quite well but for gliders in a dive this system breaks down.

Honestly I don't see why any of the navigation stuff would be part of the PID profile.  It makes sense that the throttle settings are part of the battery profile (we still need something in the configurator to indicate this) but tying more settings to the profiles makes the whole system very opaque.

**Comment by DzikuVx** [comment]
Let's see if it will work in the real case flight @MrD-RC 
I hope so :)

**Comment by MrD-RC** [comment]
I had a flight today and experienced no bad behaviour at all. The settings for the PIDFF and rates are as in the images above. 

For this flight, I had to go LOS, due to a lack of a spotter at the club field. When the profile switched, it was as if nothing had happened. But, as soon as I started manoeuvring, I could tell it was different. The turning circle was much larger with profile 2 than with profile 1. The video below shows that. The flight was longer than the video, and I switched profiles quite a few times with no issues.

https://youtu.be/1g2mHOBr0_A

 I'll make some changes to the code tomorrow. I will rename it from **PID Profile** to simply **Profile**. It makes more sense, as this covers more than just PIDFFs and/or Rates. I'll also add an ODS element, as it would be very useful for tuning, or just general flight information.

**Comment by MrD-RC** [comment]
In the last updates I
- pushed up the version of the code used in the flight test
- renamed to Profiles as opposed to PID Profiles
- added an OSD element to view the active profile.

image

This should be all good. There may be some font changes, but this can be addressed in time.  From my testing, if the profiles work fine individually, switching between them won't cause problems. I've not tried anything with different filters. So it may be worth that being tested. I fly planes, I just leave the filters on the default settings.

Making changes in flight to a profile, then switching to another profile keeps the changes. Switching between profiles uses the new values that have been set. So you can tune your profiles, switching between them. Then when you land and save it will store them for all profiles.

**Comment by MrD-RC** [comment]
> For builds where speeds is strongly correlated with throttle TPA works quite well but for gliders in a dive this system breaks down.
> 
> Honestly I don't see why any of the navigation stuff would be part of the PID profile. It makes sense that the throttle settings are part of the battery profile (we still need something in the configurator to indicate this) but tying more settings to the profiles makes the whole system very opaque.

@avsaase True, but the TPA won't work in manual. So speed based manual rates can be helpful there. Either way, it means that this is useful functionality for many different types of RC pilots.

I think the with the profiles, having things that can be effected by different tunes should be in there. I really don't get why loiter direction is in there, as that really makes no difference. But loiter radius can be effected. If you have a model set up with profile 1 at 60 degrees bank angle, the other at 45 degrees. The 60 degree will be able to fly a smaller loiter radius than the 45 degree. Should this be only set for the shallowest bank, or should it be configurable in the config profile?

File: src/main/programming/logic_condition.c [hunk 2]
```diff
@@ -333,6 +334,21 @@ static int logicConditionCompute(
             }
             break;
 
+        case LOGIC_CONDITION_SET_PROFILE:
+            operandA--;
+            if ( getConfigProfile() != operandA  && (operandA >= 0 && operandA < MAX_PROFILE_COUNT)) {
+                bool profileChanged = false;
+                if (setConfigProfile(operandA)) {
+                    pidInit();
+                    pidInitFilters();
```
Comment (DzikuVx) [inline src/main/programming/logic_condition.c:343]: @MrD-RC Just realized this code should call `schedulePidGainsUpdate();` as well

**Review by DzikuVx — CHANGES_REQUESTED** [review]
One final request

File: src/main/programming/logic_condition.c (hunk 2 above)
Comment (MrD-RC) [reply src/main/programming/logic_condition.c:343]: Added

**Review by DzikuVx — APPROVED** [review]

**Comment by stuwe911** [comment]
Darren,
quick feedback from todays test. Autotune does work when changing profiles in flight. however the OSD does seem to have a issue. it never shows any changing PID's during Autotune in the active profile. once you leave that profile lets say form 1 to 2 and then back to 1 is shows different numbers but only after a while like 1min. I thought this might be after saving? (I never manually saved up in the air). after comparing the PID settings before and after the flight it did change all settings as expected what Autotune would do so the numbers are "correct" in the FC after the flight. However comparing these new numbers with the video it shows that the OSD never showed the "real" numbers set by the FC. total fantasy numbers...

I did use the first version that you gave me. no update to the latest one. so maybe that feedback is obsolete but in case not, i hope it helps. seeing the actual changes during Autotune is kinda helpful.

Other question: do you have any idea if its possible to change the behavior of Autotune? like setting its strength? Or is it just what it is? Since i'm never happy with the result of AT i only use it for the FF and do the PID manually. with now 3 profiles it would be very handy to set AT to a level of your taste. 
cheers mate

**Comment by MrD-RC** [comment]
@stuwe911 it may well be worth trying that with the new version. It looks like the schedulePidGainsUpdate call may fix that. Flying has been a total bust for me these last 2 weeks due to the weather.

With regards to the tuning. I believe the plan is to just have autotune set rates and FF for the time being. Then other processes are being worked on to tune the P, I, and D. https://github.com/iNavFlight/inav/pull/7461 @avsaase and @b14ckyy are looking at that at the moment.

**Comment by stuwe911** [comment]
@MrD-RC i was able to do a quick test flight today but it was freezing.. flashed the latest FW as directed, diff'ed all the settings 1:1 from the old setting and went up. the good news is the OSD is now showing the changes.
the bad news is - the profile change no longer works anymore. I've double checked. all the programming LC's are exactly the same. strangely enough he started with profile 3 - which is in my case the "high speed" one. usually starts with profile 1. not sure if it's just my setup or what but i assure you i flew with the same settings as when i did the successful speed PID changing flight. just so you know.
cheers
stuwe

**Comment by MrD-RC** [comment]
@stuwe911 I've just checked the current master with a basic switch set up to change the profile, and it seems to be working. I've re-uploaded to the dropbox with that master that I've tested with.

https://youtu.be/zwijYv2yKC8

**Comment by stuwe911** [comment]
> 
> 
> @stuwe911 I've just checked the current master with a basic switch set up to change the profile, and it seems to be working. I've re-uploaded to the dropbox with that master that I've tested with.
> 
> https://youtu.be/zwijYv2yKC8

ok thank you. i will test on my end with the freshly uploaded master and let you know how it went

**Comment by stuwe911** [comment]
@MrD-RC did some tests today and all worked just well! I found the issue from the past. the operation ID was in the first version set to 41 and now it's at 42. this resulted in my code to set loiter_radius_override instead of set_profile. no wonder it was not working :) but all good now. it also works in automated WP missions as the test showed today! excellent work! Thanks

**Comment by MrD-RC** [comment]
Awesome @stuwe911 great to hear. Yeah, sorry I should have mentioned that with the new files. Brain fart, so I didn’t think about copying in the logic from a diff.

**Comment by stuwe911** [comment]
@MrD-RC no worries Darren!! it was my mistake. I only glanced over the programming tab but did not look properly in detail when going up with v2.
A little addition that i found: after rebooting, the profile selection becomes active. regardless if done via speed or switch. now if you're not flying or have the remote on it will not allow you anymore to check the content of profile 2 and 3 (depending on the setup) you then first have to go in, disable the LC, change PID's and NOT FORGET to enable the LC again. I understand this is how it works but it's kinda not so ideal. cool would be if the flight related LC's are only active when actually flying... but i guess that's a big change. once you know it's ok but at first i was a bit puzzled why the configurator keeps jumping away from profile 2 and 3. however no biggie!
cheers mate

**Comment by MrD-RC** [comment]
@stuwe911 it would be great if there was something to determine if we are flying. But there's nothing in iNav for that at the moment. There is a big discussion on it somewhere (Found it https://github.com/iNavFlight/inav/issues/2331). But nothing has happened.

However, as yours is based on speed. You could always add logic in the programming. So if the speed is less than 25cm a second, don't change the profiles at all. This would allow you to change profiles in configurator without having to worry about them switching because of the logic.

**Comment by stuwe911** [comment]
@MrD-RC yeah true - i remember a discussion years ago when i requested a change that after a crash the battery would not get drained when the plane tries RTH and this failed due to the same reason. kinda strange... a FC software that does not know if it's flying or not... :)
excellent idea! I might just do that. Set a logic for profile 1 to be activated if speed>25cm!! brilliant brain you are!! I guess that's why you're a developer and I'm only a pilot :)

**Comment by MrD-RC** [comment]
Nah, that’ll still cause you issues. Just have a switch for > 25cm/s. If that’s not active, then don’t allow the other profile changing switches to work. Otherwise you’ll just be locked to profile 1 in configurator. 

You just want it so that when it’s on the bench, the profiles are left alone. But when you throw it, it switches to your low speed profile. 25cm/s should be fine. It’s less than 1Km/h. Walking speed is 5Km/h.

**Comment by MrD-RC** [comment]
@stuwe911 I've added some examples to the profiles document. The 2nd is the speed switching with on-workbench detection 😆 How does this compare to your LCs?

https://github.com/iNavFlight/inav/blob/master/docs/Profiles.md

**Comment by stuwe911** [comment]
that is way more clever programmed than mine! I forgot the NOR operand...

you know what is really missing in the configurator? a comments tab to make
notes and descriptions for not every day programmer that they remember a
year later or if it really gets complicated. I'll copy your code going
forward. i sure it works flawlessly!

**Comment by stuwe911** [comment]
PID speed control

**Pull request merged**